scheduler/pkg/klcpermit/workflow_manager.go

```diff
@@ -54,73 +56,95 @@ type Manager interface {
 type WorkloadManager struct {
 	dynamicClient dynamic.Interface
 	Tracer        trace.Tracer
+	bindCRDSpan   map[string]trace.Span
 }
 
 func NewWorkloadManager(d dynamic.Interface) *WorkloadManager {
 	sMgr := &WorkloadManager{
 		dynamicClient: d,
 		Tracer:        otel.Tracer("keptn/scheduler"),
+		bindCRDSpan:   make(map[string]trace.Span, 100),
 	}
 	return sMgr
 }
 
-var bindCRDSpan = make(map[string]trace.Span, 100)
+func (sMgr *WorkloadManager) ObserveWorkloadForPod(ctx context.Context, handler framework.WaitingPod, pod *corev1.Pod) {
+	factory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(sMgr.dynamicClient, 0, pod.GetNamespace(), nil)
 
-func (sMgr *WorkloadManager) Permit(ctx context.Context, pod *corev1.Pod) Status {
-	//List workloadInstance run CRDs
-	name := getCRDName(pod)
-	crd, err := sMgr.GetCRD(ctx, pod.Namespace, name)
+	informer := factory.ForResource(workloadInstanceResource)
 
-	if err != nil {
-		klog.Infof("[Keptn Permit Plugin] could not find workloadInstance crd %s, err:%s", name, err.Error())
-		return WorkloadInstanceNotFound
-	}
+	sMgr.startWatching(ctx, informer.Informer(), pod, handler)
+}
+
+func (sMgr *WorkloadManager) startWatching(ctx context.Context, s cache.SharedIndexInformer, pod *corev1.Pod, handler framework.WaitingPod) {
+	stopCh := make(chan struct{})
+	workloadInstanceName := getCRDName(pod)
+
+	checkWorkloadInstance := func(obj interface{}) {
```

Comment:
maybe we could reduce complexity here by extracting in a separate getHandlerFunc(handler) the add and update handler logic

Reply:
this part would than be easily testable